Data/Aeson/Types/FromJSON.hs

```diff
@@ -1273,38 +1274,72 @@ instance (ProductFromJSON arity f, ProductSize f
 
 --------------------------------------------------------------------------------
 
+class FieldNames f where
+    fieldNames :: f a -> [Text] -> [Text]
+
+instance (FieldNames a, FieldNames b) => FieldNames (a :*: b) where
+    fieldNames _ =
+      fieldNames (undefined :: a x) .
+      fieldNames (undefined :: b y)
+
+instance (Selector s) => FieldNames (S1 s f) where
+    fieldNames _ = (pack (selName (undefined :: M1 _i s _f _p)) :)
+
 class RecordFromJSON arity f where
     recordParseJSON
+        :: Bool :* ConName :* TypeName :* Options :* FromArgs arity a
+        -> Object -> Parser (f a)
+
+instance ( FieldNames f
+         , RecordFromJSON' arity f
+         ) => RecordFromJSON arity f where
+    recordParseJSON (fromTaggedSum :* p@(cname :* tname :* opts :* _)) =
+        \obj -> checkUnknown obj >> recordParseJSON' p obj
```

Comment:
Is there a reason for these lambdas? Binding these variables on the left of `=` seems more natural.

Reply:
I'm not really sure how Haskell optimizes. My idea is that I don't want knownFields or if statement in checkUnknown to be computed only when given obj, but rather precomputed, before the function is aware of obj. I know in some functional languages, this matters (eg: sml), which is why I put it this way. Constant propagation would probably make this a non-concern but it doesn't significantly hurt readability anyway.

Reply:
Fair enough.